File: LAb_1/Lab1_ex4/Lab1_ex4.py
'''
    
      ________________________________
                        |*******************************|
                        |*******************************|
                        |***       Vũ Huy Hoàng      ***|
                        |***         20204981        ***|
                        |***         IT_E6-05        ***|
                        |*******************************|
                        |*******************************|                
    
    '''

def cac_max_rect(Mat):
    """
    Lý do phải thêm 1 hàng 0 ở đây là do mình lấy số 0 làm mốc để tính chiều
    cao của hình chữ nhật nên mới thế( vì có th hàng đầu tiên là số 1 thì nó 
    bị ngu nên không tính được height)
    """
    list = []
    for i in range (len(Mat[0])):
        list.append(0)
    #print(list)
    Mat.insert(0,list)
    H = Mat
    '''  ____________c
        |
        |
        |
        r
        
    '''
    # caculate the height of rect
    '''# đoạn này do duyệt từ trên xuống nên là daijoubu, nó sẽ ghi đè nếu gặp 
    ô số 0 gần hơn '''
    for i in range (len(H)):
        for j in range(len(H[i])):
            if H[i][j] == 0:
                for k in range(i,len(H)):
                    if H[k][j] != 0:
                        H[k][j] = k - i
    H.pop(0)
    return H
    #print(H)

def largest_rect_solve2(Mat):
    Max_area = 0

    for k in range(len(Mat)):
        for i in range(len(Mat[k])-1):
            for j in range(i+1,len(Mat[k])):
                Max_area =max(Max_area, (j-i+1)*min(Mat[k][i:j+1]))
 
    return Max_area
    

def largest_rect(Mat):
    Max_area = 0
    #  Duyệt thuật toán từ từng hàng của ma trận bổ sung
    
    for k in range(len(Mat)):
        stack = []
        
        for i,h in enumerate(Mat[k]):
            start = i
            while stack and stack[-1][1] > h:
                index,height = stack.pop()
                Max_area = max(Max_area,height*(i-index))
                start = index
            stack.append((start,h))
        for i ,h in stack:
            Max_area = max(Max_area,h*(len(Mat[k])-i))
       

        '''for i in range(len(Mat[k])-1):
            for j in range(i+1,len(Mat[k])):
                Max_area =max(Max_area, (j-i+1)*min(Mat[k][i:j+1]))'''
 
    return Max_area


A = []
n,m = map(int,input().split())
for i in range(n):
    list = input().split()
    
    A.append([eval(j) for j in list])

'''
for i in range(len(A)):
    print(A[i])'''
#cac_max_rect(A)
B = cac_max_rect(A)
'''for i in range(len(B)):
    print(B[i])'''
for i in range(len(B)):
    print(B[i])
print(largest_rect(B))

#largest_rect(B)

File: LAb_1/Lab1_ex4/Lab1_ex4.cpp
#include<iostream>
using namespace std;
#include <vector>
#include <bits/stdc++.h>
#include<stack>
std::vector<vector<int>> cac_max_rect(std::vector<vector<int>> Mat,int r,int c){
    std::vector<int> list ;
    for (int i = 0;i<c;i++){
        list.push_back(0);
    }
    Mat.insert(Mat.begin(),list);
    //std::vector<vector<int>> H = Mat;
    for (int i = 0;i<r+1;i++){
        for (int j = 0;j<c;j++)
            {if(Mat[i][j] == 0){
                for(int k = i+1;k<r+1;k++){
                    if (Mat[k][j] != 0){
                        Mat[k][j] = k - i;
                    }
                } 
                }
        }
    }
    Mat.erase(Mat.begin());
    return Mat;
}




int solve(std::vector<vector<int>> Mat,int r,int c){
    int Max_area = 0;   
    vector<int> heights(c);
    for (int i =0;i<r;i++){
        stack<int> stk;
        for (int j = 0;j<heights.size();j++){
            if(j<c){
                heights[j] = (Mat[i][j] ==1)?heights[j]+1:0;

            }
            while(!stk.empty()&& heights[j]<heights[stk.top()]){
                int cur = stk.top();
                stk.pop();
                Max_area = max(Max_area,heights[cur] * (stk.empty()?j:(j-1-stk.top())));


            }
            stk.push(j);
        }
    }
    return Max_area;
}

int main(){
    int r,c;
    std::vector<vector<int>> Mat;
    std::vector<int> list;
    cin>>r>>c;
    for (int i = 0;i<r;i++){
        for (int j = 0;j<c;j++){
            int q;
            cin>>q;
            list.push_back(q);
        }
        Mat.push_back(list);
        list.clear();

    } 
//Mat = cac_max_rect(Mat,r,c);


/*for (int i = 0;i<r+1;i++){
        for (int j = 0;j<c;j++){
            cout<<Mat[i][j]<<" ";
        }
        cout<<endl;
    } 
*/
//cout<<"F";
int ans = solve(Mat,r,c);
cout<<solve(Mat,r,c);

    return 0;
}
